crates/wasmparser/src/validator/core.rs

```diff
@@ -730,31 +736,44 @@ impl Module {
         Ok(())
     }
 
-    pub(crate) fn matches(&self, ty1: ValType, ty2: ValType, types: &TypeList) -> bool {
-        fn is_num(ty: ValType) -> bool {
-            matches!(
-                ty,
-                ValType::I32 | ValType::I64 | ValType::F32 | ValType::F64 | ValType::Bot
-            )
-        }
-
-        fn is_vec(ty: ValType) -> bool {
-            ty == ValType::V128
-        }
-
-        fn is_ref(ty: ValType) -> bool {
-            !(is_num(ty) || is_vec(ty)) || ty == ValType::Bot
-        }
-
-        fn eq_fns(f1: &impl WasmFuncType, f2: &impl WasmFuncType) -> bool {
-            f1.len_inputs() == f2.len_inputs()
-                && f2.len_outputs() == f2.len_outputs()
-                && f1.inputs().zip(f2.inputs()).all(|(t1, t2)| t1 == t2)
-                && f1.outputs().zip(f2.outputs()).all(|(t1, t2)| t1 == t2)
+    fn eq_valtypes(&self, ty1: ValType, ty2: ValType, types: &TypeList) -> bool {
+        match (ty1, ty2) {
+            (ValType::Ref(rt1), ValType::Ref(rt2)) => {
+                rt1.nullable == rt2.nullable
+                    && match (rt1.heap_type, rt2.heap_type) {
+                        (HeapType::Func, HeapType::Func) => true,
+                        (HeapType::Extern, HeapType::Extern) => true,
+                        (HeapType::Index(n1), HeapType::Index(n2)) => {
+                            let n1 = self
+                                .func_type_at(n1, types, 0)
+                                .expect("TODO fixme bad type");
+                            let n2 = self
+                                .func_type_at(n2, types, 0)
+                                .expect("TODO fixme bad type");
+                            self.eq_fns(n1, n2, types)
+                        }
+                        (_, _) => false,
+                    }
+            }
+            _ => ty1 == ty2,
         }
+    }
+    fn eq_fns(&self, f1: &impl WasmFuncType, f2: &impl WasmFuncType, types: &TypeList) -> bool {
+        f1.len_inputs() == f2.len_inputs()
```

Comment:
I don't know Rust well enough to judge whether it is better to write code in this functional style (it is certainly nice) or whether we should write in a more imperative style, e.g. using `for`-loops.